src/loki_config.py

```diff
@@ -134,15 +134,25 @@ def _common_config(self, coordinator: Coordinator) -> Dict[str, Any]:
             "storage": {"s3": self._s3_storage_config(coordinator)},
         }
 
-    def _compactor_config(self, retention_period: int) -> Dict[str, Any]:
+    def _compactor_config(self, retention_period: int, coordinator: Coordinator) -> Dict[str, Any]:
         # Ref: https://grafana.com/docs/loki/latest/configure/#compactor
         retention_enabled = retention_period != 0
-        return {
+        config = {
             # Activate custom retention. Default is False.
             "retention_enabled": retention_enabled,
             "working_directory": COMPACTOR_DIR,
         }
 
+        # When retention is enabled, it is required to configure a store for deletion requests (ref: https://grafana.com/docs/loki/latest/operations/storage/retention/)
+        # If missing, this worker will have an error: "ERROR: invalid compactor config: compactor.delete-request-store should be configured when retention is enabled`"
+        # When retention is enabled (i.e. we want to delete logs periodically), delete requests are created and the compactor executes these.
+        # It's necessary to persist these (e.g. in S3) across restarts for example.
+        # This store is the same S3 storage that we define in the storages config
+        if retention_enabled and coordinator.s3_ready:
+            config["delete_request_store"] = "aws"
```

Comment:
- If s3 is not ready then we'd get the same invalid config we used to have before this PR. Not sure if it makes sense to do it. I.e. do we need to pass `coordinator` arg? Seems like caller should guard. If S3 isn't available I wouldn't set [`filesystem`](https://github.com/grafana/loki/blob/9d44db8d6c990d7e11fdf1a17b6fc79b84f904a1/docs/sources/get-started/quick-start/tutorial.md?plain=1#L429) I think.
- Why `aws`? Some refs with `s3`: [1](https://github.com/grafana/loki/blob/9d44db8d6c990d7e11fdf1a17b6fc79b84f904a1/tools/dev/loki-tsdb-storage-s3/config/loki.yaml#L15), [2](https://github.com/grafana/loki/blob/9d44db8d6c990d7e11fdf1a17b6fc79b84f904a1/production/nomad/loki/config.yml#L59). Can you please confirm manually that it works?